net/netlink/netlink_netfilter.c

```diff
@@ -0,0 +1,707 @@
+ * +------+--------------+-----------------+
+ * | attr | CTA_TUPLE_IP | CTA_TUPLE_PROTO |
+ * +------+--------------+-----------------+
+ */
+
+/* CTA_TUPLE_IP definitions */
+
+struct conntrack_tuple_ipv4_s
+{
+  struct nfattr attr;
+  struct nfnl_ipv4addr_s src;
+  struct nfnl_ipv4addr_s dst;
+};
+
+struct conntrack_tuple_ipv6_s
+{
+  struct nfattr attr;
+  struct nfnl_ipv6addr_s src;
+  struct nfnl_ipv6addr_s dst;
+};
+
+/* CTA_TUPLE_PROTO definitions */
+
+struct conntrack_tuple_tcpudp_s
+{
+  struct nfattr attr;
+  struct nfnl_attr_u8_s  proto;
+  struct nfnl_attr_u16_s sport;
+  struct nfnl_attr_u16_s dport;
+};
+
+struct conntrack_tuple_icmp_s
+{
+  struct nfattr attr;
+  struct nfnl_attr_u8_s  proto;
+  struct nfnl_attr_u16_s id;
+  struct nfnl_attr_u8_s  type;
+  struct nfnl_attr_u8_s  code;
+};
+
+/* Struct of a conntrack response
+ * +-----+-----+-----------------+----------------+
+ * | hdr | msg | tuple of origin | tuple of reply |
+ * +-----+-----+-----------------+----------------+
+ */
+
+struct conntrack_recvfrom_response_s
+{
+  struct nlmsghdr hdr;
+  struct nfgenmsg msg;
+  uint8_t         data[1];
+};
+
+#define SIZEOF_CTNL_RECVFROM_RESPONSE_S(n) \
+  (sizeof(struct conntrack_recvfrom_response_s) + (n) - 1)
+
+struct conntrack_recvfrom_rsplist_s
+{
+  sq_entry_t flink;
+  struct conntrack_recvfrom_response_s payload;
+};
+
+#define SIZEOF_CTNL_RECVFROM_RSPLIST_S(n) \
+  (sizeof(struct conntrack_recvfrom_rsplist_s) + (n) - 1)
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: netlink_conntrack_tuple_size
+ *
+ * Description:
+ *   Get the size of a CTA_TUPLE. Struct of a conntrack tuple:
+ *     +------+--------------+-----------------+
+ *     | attr | CTA_TUPLE_IP | CTA_TUPLE_PROTO |
+ *     +------+--------------+-----------------+
+ *
+ * Input Parameters:
+ *   domain - The domain of the tuple
+ *   proto  - The protocol of the tuple
+ *
+ * Returned Value:
+ *   The size of the tuple.
+ *
+ ****************************************************************************/
+
+static ssize_t netlink_conntrack_tuple_size(uint8_t domain, uint8_t proto)
+{
+  size_t size = sizeof(struct nfattr);
+
+  switch (domain)
+    {
+      case PF_INET:
+        size += sizeof(struct conntrack_tuple_ipv4_s);
+        break;
+
+      case PF_INET6:
+        size += sizeof(struct conntrack_tuple_ipv6_s);
+        break;
+
+      default:
+        return -EINVAL;
+    }
+
+  switch (proto)
+    {
+      case IPPROTO_TCP:
+      case IPPROTO_UDP:
+        size += sizeof(struct conntrack_tuple_tcpudp_s);
+        break;
+
+      case IPPROTO_ICMP:
+      case IPPROTO_ICMP6:
+        size += sizeof(struct conntrack_tuple_icmp_s);
+        break;
+
+      default:
+        return -EINVAL;
+    }
+
+  return size;
+}
```

Comment:
```suggestion
static ssize_t netlink_conntrack_tuple_size(uint8_t domain, uint8_t proto)
{
  size_t size;

  switch (domain)
    {
      case PF_INET:
        size = sizeof(struct conntrack_tuple_ipv4_s);
        break;

      case PF_INET6:
        size = sizeof(struct conntrack_tuple_ipv6_s);
        break;

      default:
        return -EINVAL;
    }

  switch (proto)
    {
      case IPPROTO_TCP:
      case IPPROTO_UDP:
        size = sizeof(struct conntrack_tuple_tcpudp_s);
        break;

      case IPPROTO_ICMP:
      case IPPROTO_ICMP6:
        size = sizeof(struct conntrack_tuple_icmp_s);
        break;

      default:
        return -EINVAL;
    }

  return size + sizeof(struct nfattr);
}
```

Reply:
We need to get the total size of `attr + CTA_TUPLE_IP + CTA_TUPLE_PROTO`.
Also, `size_t size` on the stack still needs to be initialized as 0, and init it as `sizeof(struct nfattr)` won't be much slower.

Reply:
Yep, I lost the PROTO part.